EpikCord/client/http.py

```diff
@@ -0,0 +1,475 @@
+        channel_id: Optional[int] = None,
+        guild_id: Optional[int] = None,
+        webhook_id: Optional[int] = None,
+        webhook_token: Optional[str] = None,
+    ) -> Union[Bucket, TopLevelBucket]:
+        url = response.url
+        method = response.method
+
+        if channel_id or guild_id or webhook_id or webhook_token:
+            bucket = TopLevelBucket(
+                channel_id=channel_id,
+                guild_id=guild_id,
+                webhook_id=webhook_id,
+                webhook_token=webhook_token,
+                )
+
+        else:
+
+            bucket = Bucket(bucket_hash=response.headers["X-RateLimit-Bucket"])
+            if bucket in self.buckets.values():
+                self.buckets[f"{method}:{url}"] = list(self.buckets.values())[
+                    list(self.buckets.values()).index(bucket)
+                ]
+                bucket = self.buckets[f"{method}:{url}"]
+            self.buckets[f"{method}:{url}"] = bucket
+
+        return bucket
+
+    async def extract_content(self, response: aiohttp.ClientResponse) -> Dict[str, Any]:
+        if response.headers["Content-Type"] == "application/json":
+            data = await response.json()
+        else:
+            data = {}
+        return data
+
+    async def handle_exhausted_bucket(
+        self, response: aiohttp.ClientResponse, bucket: Union[Bucket, TopLevelBucket]
+    ):
+        bucket.clear()
+        await asyncio.sleep(int(response.headers["X-RateLimit-Reset-After"]))
+        bucket.set()
+
+    async def handle_ratelimit(
+        self, data: Dict[str, Any], bucket: Union[Bucket, TopLevelBucket]
+    ):
+        bucket.clear()
+
+        if data["global"]:
+            self.global_event.clear()
+
+        await asyncio.sleep(data["retry_after"])
+
+        self.global_event.set()
+        bucket.set()
+
+    def clean_url(self, url: str) -> str:
+        if url.startswith("/"):
+            url = f"https://discord.com/api/v{self.version}{url}"
+        else:
+            url = f"https://discord.com/api/v{self.version}/{url}"
+
+        if url.endswith("/"):
+            url = url[:-1]
+
+        return url
+
+    @staticmethod
+    async def log_request(res, body: Optional[dict] = None):
+        messages = [
+            f"Sent a {res.method} to {res.url} " f"and got a {res.status} response. ",
+            f"Content-Type: {res.headers['Content-Type']} ",
+        ]
+
+        if body:
+            messages.append(f"Sent body: {body} ")
+
+        if h := dict(res.request_info.headers):
+            messages.append(f"Sent headers: {h} ")
+
+        if h := dict(res.headers):
+            messages.append(f"Received headers: {h} ")
+
+        try:
+            messages.append(f"Received body: {await res.json()} ")
+
+        finally:
+            logger.debug("".join(messages))
+
+    def setup_kwargs(
+        self, kwargs, *, files: Optional[List[File]] = None, json: Optional[Dict] = None
+    ) -> None:
```

Comment:
Wouldn't it make sense for `setup_kwargs` to return the updated kwargs like in `clean_url`?
Also this should be a static method.

Reply:
But that isn't *needed*, it modifies the `kwargs` passed.